clang/test/SemaSYCL/builtin-sycl-kernel-name.cpp

```diff
@@ -0,0 +1,44 @@
+// RUN: %clang_cc1 -fsycl-is-host -fsyntax-only -verify %s
+
+class kernel_name_1;
+class kernel_name_2;
+class kernel_name_3;
+typedef kernel_name_3 kernel_name_TD;
+
+template<typename KN>
+struct kernel_id_1 {
+  using type = KN;
+};
+
+struct kernel_id_2 {
+  using type = kernel_name_2;
+};
+
+struct kernel_id_3 {
+  using invalid_name = kernel_name_2;
+};
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel_entry_point(name))) void kernel_single_task(const Func kernelFunc) {
+  kernelFunc();
+}
+
+struct SYCLKernel {
+  int m;
+  public:
+  void operator()() const {}
+};
+
+void test() {
+  SYCLKernel Obj;
+  kernel_single_task<kernel_name_1>(Obj);
+  kernel_single_task<kernel_name_2>(Obj);
+  kernel_single_task<kernel_name_TD>(Obj);
+  const char* test1 = __builtin_sycl_kernel_name(kernel_id_1<kernel_name_1>()); // Valid
+  const char* test2 = __builtin_sycl_kernel_name(kernel_id_1<kernel_name_TD>()); // Valid
+  const char* test3 = __builtin_sycl_kernel_name(kernel_id_2()); // Valid
+  const char* test4 = __builtin_sycl_kernel_name(kernel_id_3()); // expected-error {{invalid argument; argument must be a class or struct type with a member type alias named 'type'}}
+  const char* test5 = __builtin_sycl_kernel_name("str"); // expected-error {{invalid argument; argument must be a class or struct type with a member type alias named 'type'}}
+  const char* test6 = __builtin_sycl_kernel_name(kernel_id_2(), kernel_id_2()); // expected-error {{builtin requires exactly 1 argument}}
+}
+
```

Comment:
Additional test suggestions:

Test invocation in constant evaluation:
```
struct constexpr_kernel_name1;
void test_constexpr() {
  kernel_single_task<constexpr_kernel_name1>([]{});
  constexpr const char* test1 = __builtin_sycl_kernel_name(kernel_id_1<constexpr_kernel_name1>()); // Valid
}
```

Test that a non-constant argument produces an error.
```
struct nonconst_kernel_name1;
struct nonconst_kernel_name2;
void test() {
  kernel_single_task<nonconst_kernel_name1>([]{});
  kernel_single_task<nonconst_kernel_name2>([]{});
  constexpr kernel_id_1<nonconst_kernel_name1> kn1;
  (void)__builtin_sycl_kernel_name(kn1); // Valid
  kernel_id_1<nonconst_kernel_name2> kn2;
  (void)__builtin_sycl_kernel_name(kn2); // expected-error {{FIXME}}
}
```

Reply:
Additional cases to exercise:
- Calling the builtin in a constexpr context before a call to `kernel_single_task()`; ensure an error is produced.
- Calling the builtin before a call to `kernel_single_task()`; this should work by accessing data in a symbol declaration for which a symbol definition is emitted sometime after the call to `kernel_single_task()` is observed (perhaps at the end of the translation unit contingent on use of the symbol).
- Calling the builtin with no corresponding call to `kernel_single_task()`; ensure an error is produced at the end of the translation unit. This case is similar to the previous case; the error occurs because the symbol definition is never emitted.